### ISSUE TITLE
Rethink language selector again 😀

### ISSUE THREAD

**Issue by schalkneethling** [mdn/yari#1486]
While working on aligning all of the pieces of the page on the various viewport sizes I realized we have a bit of a UI/UX problem on mobile due to the components we have in our overall page header section. This is best described via a screenshot of a reference page on mobile:

![Screenshot 2020-10-19 at 10 16 10](https://user-images.githubusercontent.com/10350960/96419592-a43bb180-11f4-11eb-8f8a-3abb6b361d5e.png)

As you can see from the above, not only does it feel cluttered/overwhelming, it pushes the actual content of the page further and further down. I, therefore, started rethinking how we surface a language selector.

First, a general question. Yari 1 will ship with translations but, we will link to frozen translations from the language selector until we relaunch translations. Secondly, translations will have a banner that indicates that the translation is frozen and, will offer a link back to the up to date English version. Third, there will be a period of time, 4-6 months perhaps, between Yari 1 and the relaunch of translations where the frozen translations for some pages will get out of date so, do we want to actively provide the means to link to them? 

With all of that, should we perhaps consider kicking this can down the road and decide/work on its implementtion after the initial launch?

With that said and out of the way 😀 and assuming you are still reading, I do have a possible proposal we can implement in time for Yari 1. Here is my thinking around this problem.

1. Users get to translated content, as most content, from an external search engine that returned results in their native/browser/OS language. i.e. very few people change their language while just generally browsing MDN (I know, except changing it to English, see point 2 below 😀)
2. The most asked for feature is to have a setting to always display MDN in English. This then also suggests that should a user want to change the language, they are going to want to switch to English.
3. This is not a commonly used part of MDN and as such, should definitely not be so prominent and take up so much of our screen real estate.

With all of that in mind, I broke this down into the following statement:

> Provide the ability to change the current language of an MDN page in an easy, accessible, non-obtrusive manner. Make the most common use case quickly accessible.

From a UI perspective I see the above implemented as follows:

### Mobile

![Screenshot 2020-10-19 at 19 16 15](https://user-images.githubusercontent.com/10350960/96489537-d9b9bc80-123f-11eb-848e-72bf29ef0f4f.png)

### Tablet

![Screenshot 2020-10-19 at 19 16 28](https://user-images.githubusercontent.com/10350960/96489572-e4745180-123f-11eb-86a9-832028b6d37e.png)

### Desktop

![Screenshot 2020-10-19 at 19 16 43](https://user-images.githubusercontent.com/10350960/96489608-ef2ee680-123f-11eb-9565-cf8e51a277aa.png)

Couple of notes about the above:

1. The icon used is the universal language icon for selecting a different language: http://www.languageicon.org/
2. But, for accessibility the heading "Change language" is only visually hidden on mobile and tablet and,
3. The icon is a link and will have `title` and `aria-labelledby` attributes that clearly describes the links intent i.e. "Select your preferred language"
4. The `English` link, will only be shown if you are no already on an English page and as such, change to the English version when shown
5. The language icon will link to the language selector in the footer as can be seen in the screenshot below:

![Screenshot 2020-10-19 at 19 46 32](https://user-images.githubusercontent.com/10350960/96492378-d9232500-1243-11eb-9488-c43e2997a0c7.png)

I would really appreciate **everyone's** input here. Thanks!

**Comment by Ryuno-Ki** [comment]
I'm one of the person, who often switch to English (from German, my mother tongue).
If I get some time at hand, I come back that day and update the translation of the article.

To me, a position in the footer (like on AMO), would be totally okay.
I think, it's a position (= the footer), where people came to expect looking for a language switch (or in the header like now).
You could run some tests to be sure.

If you automatically translate the content, flag it as such. The quality is mixed. Is there a way to measure the recentness of the translation? (I think, I saw an index page on MDN somewhere which lists the status of translations … a link to it would be handy here to look at numbers?).
Perhaps it's not worth to link to translated versions but default to the English one for the time of deploying Yari.
Or point to a static copy of Kuma-MDN somewhere?

**Comment by jess-cook03** [comment]
@schalkneethling 
I think if we are looking for a solution to implement in time for Yari 1..... this could work... that being said, I have some thoughts on the proposed solution:

- I think the size and placement in the header area makes sense (via your #3 point above) -- visually the size of the icon should be the same size as the other iconography that is in-line with the text. Right now its looking a little clunky.
The universal language icon, totally makes sense, in fact mozilla uses the two and as a simplified variation within the protocol library that could used... I am suggesting this because at such a small size this current icon actually becomes quite busy and hard to understand because of its intricate detail.
Here is a screenshot of the protocol language icon:

- The icon + | English ..... the actual | or pipe or line that is currently being used I am not sure needs to be there, I get the logic for wanting to separate the two items but in reality they go together and I see that based on your proposed functionality the 'English' text will only appear if you are not already on an English page... so the state you are showing in the screenshots above --- how often will the two appear together?

I think the possibility of having both the icon + and text link functioning the same way could be worth exploring rather than having the icon function one way and the text link the other (ie I'm suggesting clicking either/ both takes you to the language selector...?)

Or maybe its a situation where we only show the 'english' text link when needed (not with the icon... user clicks and it switches page to English) and vice versa show the icon by itself in all other states and clicking it brings up the language selector .... ?  More of a one or the other scenario based on context.

Thinking about this acting as essentially an anchor link from the header area to footer.... I also question if this is needed or if its the right move (per the comment above).... its a good point.

> To me, a position in the footer (like on AMO), would be totally okay.
I think, it's a position (= the footer), where people came to expect looking for a language switch (or in the header like now).
You could run some tests to be sure.

All other Mozilla web spaces keep the language selector in the footer.... including on support.mozilla.org ... I think it is very much an expected user behavior to go to the footer anyways to select language.

I agree I would think some testing would be useful here to see how it performs vrs. just having the traditional language selector in the footer.

With that being said:
I think potentially tweaking the current language dropdown component that lives in the footer would be most ideal regardless of implementing the header link suggested above...
Taking the step of the 'select' button away and simplifying the interaction ....
This could be refined to: 
1) open the dropdown 2) select language and page refreshes automatically upon selection
Can refer to the footers on mozilla.org or support.mozilla.org

I hope this helps, I think on Wednesday we can discuss this further and make a decision we want to move forward with! 👍

**Comment by peterbe** [comment]
> All other Mozilla web spaces keep the language selector in the footer.... including on support.mozilla.org ... I think it is very much an expected user behavior to go to the footer anyways to select language.

One thing we've seen many times is that people google search for something (e.g. "array prototype for each") and, because they *live* in France, the French Google sends them to https://developer.mozilla.org/fr/docs/Web/JavaScript/Reference/Objets_globaux/Array/forEach but the user codes in English and prefers the latest-and-greatest version in English so they switch over to English. 

Considering that this comes up again and again, I think it's a good argument to keep the language switching choice near the "top" of the document.

**Comment by peterbe** [comment]
> First, a general question. Yari 1 will ship with translations but, we will link to frozen translations from the language selector until we relaunch translations. Secondly, translations will have a banner that indicates that the translation is frozen and, will offer a link back to the up to date English version. Third, there will be a period of time, 4-6 months perhaps, between Yari 1 and the relaunch of translations where the frozen translations for some pages will get out of date so, do we want to actively provide the means to link to them?

Accurate. 
Also, unlike today's Kuma, when you're on the French document translation, we don't have a list of the other translations. Only, and always, to English. 

So if you're on a Japanese page, does it even make sense to offer any choice? It can be as simple as:

```html
<a href="/en-US/docs/$CURRENT_SLUG">
  <svg path="language-icon"/>
  English
</a>
```

I.e. no choices or widgets or anything fancy. Just a link that says "English". 

Next, if you're on an English page, we might have a list of *other* locales. In that case, I would be happy if it just said "Change language <svg path=language-icon>" and when you click it, the (select) widget appears. 

Actually, I'm still a bit confused about the suggested design above. It looks like it's like this:

```html
<img src="language-icon.svg"> | English
```
Is it a hyperlink? What does it go to?

**Comment by schalkneethling** [comment]
>     * visually the size of the icon should be the same size as the other iconography that is in-line with the text. Right now its looking a little clunky.
>       The universal language icon, totally makes sense, in fact mozilla uses the two and as a simplified variation within the protocol library that could used... I am suggesting this because at such a small size this current icon actually becomes quite busy and hard to understand because of its intricate detail.
>       Here is a screenshot of the protocol language icon:

100% agree.

> so the state you are showing in the screenshots above --- how often will the two appear together?

As Peter also explained, the most common scenario for this would be someone doing a search on say Google, Google returns an MDN result but, because the user's locale is set to say Franch, the link points to the French version of the page.

In those cases, we will have the icon but also include the `English` link. The main reason for including the `English` link directly is because generally, that is the most common option users will choose. Having the link means they can switch to English with one click instead of clicking the icon, which takes them to the footer where they then have to choose English from the `select` element, and then click the `Select` button. Basically a little affordance for the most common use case.

> Thinking about this acting as essentially an anchor link from the header area to footer.... I also question if this is needed or if its the right move (per the comment above).... its a good point.

For sure, this is definitely another option. i.e. not have anything linking to it but just have the functionality be present in the footer.

> Taking the step of the 'select' button away and simplifying the interaction ....

I would not recommend this. We specifically changed our language selector to be this way for accessibility reasons. This ensures a predictable interaction model for both users of assistive technologies and keyboard users irrespective of ability.

You will find more information regarding this here:
https://webaim.org/techniques/forms/controls#javascript

**Comment by schalkneethling** [comment]
> Accurate.
> Also, unlike today's Kuma, when you're on the French document translation, we don't have a list of the other translations. Only, and always, to English.
> 
> So if you're on a Japanese page, does it even make sense to offer any choice? It can be as simple as:

Oh well, that changes things a lot. So, if you are on say, French, the only option would be to switch to English and only then could you for example switch to German? We will not offer this on the French page?

> Next, if you're on an English page, we might have a list of other locales. In that case, I would be happy if it just said "Change language " and when you click it, the (select) widget appears.

Indeed, in that scenario that would make complete sense.

> Is it a hyperlink? What does it go to?

Hard to see just from the static screenshot but yes, the language icon would be a hyperlink that would link to the `select` element in the footer, from where you can change your language if you did not simply want to switch to English.

**Comment by schalkneethling** [comment]
I have updated the layout with the new icon. I also forgot to include the Figma link:
https://www.figma.com/file/YYVJ8uqG8UFBBvQhdI7fxR/MDN-Web-Docs?node-id=344%3A0

**Comment by peterbe** [comment]
I know mobile share isn't huge but I think we really need something because this isn't looking great. Perhaps before we come up with the perfect solution, we can maybe aim for a hybrid fix to tie us over. 
![IMG_B5AE1A146A7C-1](https://user-images.githubusercontent.com/26739/101804556-b55fbb00-3adf-11eb-832f-a03ca7f17198.jpeg)

**Comment by mattiapontonio** [comment]
We should think about making it more standard form. In this way some issues should be resolved and perhaps new ones on global form styles should be addressed.

![Screenshot_20201210-182231.png](https://user-images.githubusercontent.com/44380480/101807191-1eadf100-3b15-11eb-8252-21e79a3f199a.png)

https://developer.mozilla.org/en-US/docs/Learn/Forms/How_to_structure_a_web_form

**Comment by schalkneethling** [comment]
@peterbe I would like to tackle this one as one of the first tasks post-launch. It really is not a great user experience on mobile with this stuck in the header along with all the other UI pieces. The tricky thing here is that the `Footer` component would need access to the `document` as it contains the neeed information about translations.

Currently we load than inside the `Document` component and so, there is no quick/simple way to get that info the the `Footer`. Well, I mean, if we break some DRY principles we could be moving the `Footer` component into the `Document`, but perhaps that is not something we want to do.

I reckon having the languages selector only available on document pages is fine and so, we do not have to be concerned with other pages such as the landing page, signup, profile etc. Perhaps that helps, perhaps it makes not real difference.

Anyhow I feel that what we landed on in https://github.com/mdn/yari/issues/1486#issuecomment-712722153 would be a much better experience for everyone, but I am going to need your assistance on how to best structure the React pieces to enable the layout change.

**Comment by mattiapontonio** [comment]
I haven't seen the code at the moment but I know the `footer` is part of the `document`. Or at least by the `body` that is part of the `document`.

**Comment by mattiapontonio** [comment]
- https://github.com/mdn/yari/issues/2046

**Comment by mnme** [comment]
![Screenshot from 2020-12-16 13-24-03](https://user-images.githubusercontent.com/2905118/102348382-20a70280-3fa2-11eb-81bc-865abdd0fba8.png)

Just a heads up: The language selector is currently unaligned for me on Firefox 84 / Arch Linux. Instead of opening a new issue I'm just leaving this comment here, I hope that's okay.

**Comment by peterbe** [comment]
> I'm just leaving this comment here, I hope that's okay.

Very OK. 

Besides, I think @schalkneethling is planning to move the language dropdown form thingy to the bottom of the page. Perhaps you can come back and check that it's properly aligned there when it gets there.

**Comment by peterbe** [comment]
@schalkneethling What's the status of this now?

**Comment by schalkneethling** [comment]
> @schalkneethling What's the status of this now?

It is on my to-do for this week. 👍

**Comment by schalkneethling** [comment]
![](https://github.trello.services/images/mini-trello-icon.png) [In content callouts](https://trello.com/c/TDDNJ6L1/16-in-content-callouts)

**Comment by mnme** [comment]
> > I'm just leaving this comment here, I hope that's okay.
> 
> Very OK.
> 
> Besides, I think @schalkneethling is planning to move the language dropdown form thingy to the bottom of the page. Perhaps you can come back and check that it's properly aligned there when it gets there.

I just remembered this issue and lo and behold, the alignment doesn't seem to be fixed (Arch Linux / Firefox 86.0.1).

![Screenshot from 2021-03-22 16-50-06](https://user-images.githubusercontent.com/2905118/112018436-d5673780-8b2e-11eb-9417-48d85751e0bd.png)

**Comment by peterbe** [comment]
> the alignment doesn't seem to be fixed

That's a bummer. Would you mind filing a fresh new issue about that. 
Also if you think you can figure out a solution we'd be delighted to guide you to how you can create a PR and stuff.

**Comment by schalkneethling** [comment]
> > > I'm just leaving this comment here, I hope that's okay.
> > 
> > 
> > Very OK.
> > Besides, I think @schalkneethling is planning to move the language dropdown form thingy to the bottom of the page. Perhaps you can come back and check that it's properly aligned there when it gets there.
> 
> I just remembered this issue and lo and behold, the alignment doesn't seem to be fixed (Arch Linux / Firefox 86.0.1).
> 
> ![Screenshot from 2021-03-22 16-50-06](https://user-images.githubusercontent.com/2905118/112018436-d5673780-8b2e-11eb-9417-48d85751e0bd.png)

Thanks for reporting. I opened an issue for this. I am going to style our select elements better now that we have settles on a solution for this component. Along with this, I will also address the issue here.
